straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,316 @@ def find_n_competing(peaks, windows, fraction):
+
+        # 6. Set time and endtime for peaks
+        result['time'] = current_peak['time']
+        result['endtime'] = strax.endtime(current_peak)
+        return result
+
+    @staticmethod
+    def getsigma(sigma, baseline, s2):
+        # The parameter of HalfCauchy, which is a function of S2 area
+        return sigma / np.sqrt(s2) + baseline
 
     @staticmethod
     @numba.njit
-    def compute_shadow(peaks, pre_s2_peaks, touching_windows, exponent, res):
+    def peaks_shadow(peaks, pre_peaks, touching_windows, exponent, result, pos_corr, sigmas=None):
         """
-        For each peak in peaks, check if there is a shadow-casting S2 peak
+        For each peak in peaks, check if there is a shadow-casting peak
         and check if it casts the largest shadow
         """
+        for p_i, (p_a, sigma) in enumerate(zip(peaks, sigmas)):
+            indices = touching_windows[p_i]
+            for idx in range(indices[0], indices[1]):
+                s_a = pre_peaks[idx]
+                dt = p_a['center_time'] - s_a['center_time']
+                if dt <= 0:
+                    continue
+                # First we record the time difference to the nearest previous peak
+                if dt < result['nearest_dt'][p_i]:
+                    result['nearest_dt'][p_i] = dt
+                # Calculate time shadow
+                new_shadow = s_a['area'] * dt ** exponent
+                if pos_corr:
+                    # Calculate position shadow which is time shadow with a HalfCauchy PDF multiplier
+                    distance = np.sqrt((p_a['x'] - s_a['x']) ** 2 + (p_a['y'] - s_a['y']) ** 2)
+                    distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+                    new_shadow *= 2 / (np.pi * sigma * (1 + (distance / sigma) ** 2))
+                # Only the previous peak with largest shadow is recorded
+                if new_shadow > result['shadow'][p_i]:
+                    result['shadow'][p_i] = new_shadow
+                    result['pre_x'][p_i] = s_a['x']
+                    result['pre_y'][p_i] = s_a['y']
+                    result['pre_dt'][p_i] = p_a['center_time'] - s_a['center_time']
+
+
+@export
+@strax.takes_config(
+    strax.Option(name='ambience_divide_time', default=False,
+                 help='Whether to divide area by time'),
+    strax.Option(name='ambience_divide_space', default=False,
+                 help='Whether to divide area by space'),
+    strax.Option(name='time_window_ambience', default=int(2e6),
+                 help='Search for ambience in this time window [ns]'),
+    strax.Option(name='ambience_area_parameters', default=(5, 60, 60), type=(list, tuple),
+                 help='The upper limit of S0, S1, S2 area to be counted'),
+    strax.Option(name='ambience_radius', default=6.7,
+                 help='Search for ambience in this radius [cm]'),
```

Comment:
Not like that. When calculating `n_sx_before`, `radius` is set as -1. In that case we do not require `r <= radius` because `r <= radius` is for `n_s2_near`.